### PR TITLE
Dynamic view sizing [dart:ui]

### PR DESCRIPTION
**Pull request by goderbauer**
Towards https://github.com/flutter/flutter/issues/134501.

This PR makes the following changes to the public dart:ui API:

* It adds the `FlutterView.pysicalConstraints` property that describes max and min width and height for a view. The framework is allowed to size the `FlutterView` to any `Size` that meets these constraints.
* It adds an optional `size` argument to `FlutterView.render`. The framework provides the chosen `Size` that meets the aforementioned constraints to the `render` method. If the `FlutterView.pysicalConstraints` are tight (minHeight == maxHeight and minWidth == maxWidth) the argument is optional to remain backwards compatible. In all other cases, a `Size` must be provided.
* It adds a `ViewConstraints` class, which is basically the `dart:ui` version of `BoxConstraints` (This is similar to how we have `ViewPadding` in dart:ui to mirror `EdgeInsets` from the framework). It describes the constraints of a `FlutterView`, i.e. it powers the `FlutterView.pysicalConstraints` property.

This change does not wire anything up to the embedders. For now, `FlutterView.pysicalConstraints` just returns tight constraints for the embedder-provided size of the view (`FlutterView.physicalSize`) and the size provided to `FlutterView.render` is ignored (after it is checked that it meets the constrains). 

This PR enables the framework to implement the new dynamic view sizing and embedders to separately expose the new functionality to their clients.

Presubmits will fail until https://github.com/flutter/flutter/pull/138565 is submitted to the framework.

**DO NOT SUBMIT until https://github.com/flutter/flutter/pull/138648 is ready.**

### CODE REVIEW

File: lib/web_ui/lib/src/engine/window.dart [hunk 1]
```diff
@@ -73,7 +73,11 @@ base class EngineFlutterView implements ui.FlutterView {
   final ViewConfiguration _viewConfiguration = const ViewConfiguration();
 
   @override
-  void render(ui.Scene scene) => platformDispatcher.render(scene, this);
+  void render(ui.Scene scene, {ui.Size? size}) {
+    assert((size != null && physicalConstraints.isSatisfiedBy(size)) || (size == null && physicalConstraints.isTight));
+    // TODO(goderbauer): Resepct the provided size when "physicalConstraints" are not allways tight. See TODO on "physicalConstraints".
```
Comment (goderbauer) [inline lib/web_ui/lib/src/engine/window.dart:78]: @ditman Wondering if you'd be up for wiring this (and the other TODO below) through the web engine.

Comment (ditman) [reply lib/web_ui/lib/src/engine/window.dart:78]: @goderbauer no problem! I think I'm going to use tight constraints as if they were a normal "size", and then do some special casing so constraints + size allow for a web-like reflow!

File: lib/web_ui/lib/src/engine/window.dart [hunk 2]
```diff
@@ -93,6 +97,10 @@ base class EngineFlutterView implements ui.FlutterView {
   late final PlatformViewMessageHandler platformViewMessageHandler =
       PlatformViewMessageHandler(platformViewsContainer: dom.platformViewsHost);
 
+  // TODO(goderbauer): Provide API to configure constraints. See also TODO in "render".
+  @override
+  ViewConstraints get physicalConstraints => ViewConstraints.tight(physicalSize);
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/window.dart:102]: These physicalConstraints must come from the native embedder, and cannot be provided by the framework, right?

Comment (goderbauer) [reply lib/web_ui/lib/src/engine/window.dart:102]: Yeah, the embedder must tell me what would be legal sizes. If you just want to have it size to content in both dimensions, you could just provide min width/height = 0 and max width/height = double.infinity.

**Comment by goderbauer** [comment]
Quick note: The framework tests are failing until https://github.com/flutter/flutter/pull/138565 is submitted to the framework to prepare for this PR.

File: lib/ui/window.dart [hunk 3]
```diff
@@ -324,25 +358,32 @@ class FlutterView {
...
+  /// If no `size` is provided, [physicalSize] is used instead.
+  ///
   /// See also:
   ///
   /// * [SchedulerBinding], the Flutter framework class which manages the
   ///   scheduling of frames.
   /// * [RendererBinding], the Flutter framework class which manages layout and
   ///   painting.
-  void render(Scene scene) {
+  void render(Scene scene, {Size? size}) {
```
Comment (goderbauer) [inline lib/ui/window.dart:374]: For some historic context: I wish we could assert here that the size passed into this method meets the physicalConstraints of the FlutterView. However, doing that is breaking a ton of things. Among other things, tests freely mock out physicalSize/physicalConstraints, which means the size the end up providing here doesn't match the real physicalConstraints. Changing the test framework to not call render in those situations (i.e. outside of onBeginFrame / onDrawFrame, https://github.com/flutter/flutter/issues/137073) breaks golden tests on web because they rely on being able to call render whenever they want.

Long story short: We decided to not assert that the size meets the constraints here in this low-level API and instead left it to a higher level to ensure that something sensible is passed here.

File: lib/ui/platform_dispatcher.dart [hunk 4]
```diff
@@ -1479,7 +1479,7 @@ class _ViewConfiguration {
   /// The pixel density of the output surface.
   final double devicePixelRatio;
 
-  /// The size requested for the view in logical pixels.
+  /// The size requested for the view in physical pixels.
```
Comment (dkwingsmt) [inline lib/ui/platform_dispatcher.dart:1482]: How did we find out this should be physical instead of logical? Is it a bug fix or a behavior change?

Comment (goderbauer) [reply lib/ui/platform_dispatcher.dart:1482]: The value was always in physical pixels, it was just documented wrong. Luckily, it is a doc comment on a private class, so the doc was never actually public.

Comment (goderbauer) [reply lib/ui/platform_dispatcher.dart:1482]: Backup evidence that this has always been in physical pixels and we just documented it wornt: This value is returned by the physicalSize getter on `FlutterView`:

https://github.com/flutter/engine/blob/dba80be699446ee6142038a9d613e367f53ffa7d/lib/ui/window.dart#L163

File: lib/web_ui/lib/src/engine/window.dart (hunk 2 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/window.dart:120]: This works. Once this gets rebased, I'll use the new JS API for the embedder to pass a Constraints object from JS into Flutter.

File: lib/ui/window.dart (hunk 3 above)
Comment (ditman) [reply lib/ui/window.dart:374]: @goderbauer should we have an issue so this can be revisited/worked on by our future selves?

File: lib/ui/platform_dispatcher.dart (hunk 4 above)
Comment (ditman) [reply lib/ui/platform_dispatcher.dart:1482]: I've found Physical vs Logical pixels is *painful* in the web, especially if one trusts the docs :P

File: lib/ui/window.dart [hunk 5]
```diff
@@ -138,8 +138,42 @@ class FlutterView {
...
-  /// will be drawn on the screen, in physical pixels.
+  /// The sizing constraints in physical pixels for this view.
+  ///
+  /// The view can take on any [Size] that fulfills these constraints. These
+  /// constraints are typically used by an UI framework as the input for its
+  /// layout algorithm to determine an approrpiate size for the view. To size
+  /// the view, the selected size must be provided to the [render] method and it
+  /// must satisfy the constraints.
+  ///
+  /// When this changes, [PlatformDispatcher.onMetricsChanged] is called.
```
Comment (ditman) [inline lib/ui/window.dart:149]: > `/// When this changes, [PlatformDispatcher...] is called.`

This bit of the documentation states that this may change, but there's not a setter for `physicalConstraints`. Should these be re-configurable after view creation? Do we need a new native API to `setPhysicalConstraints`?

File: lib/ui/window.dart [hunk 6]
```diff
@@ -138,8 +138,42 @@ class FlutterView {
...
+  ///
+  /// At startup, the constraints for the view may not be known before Dart code
+  /// runs. If this value is observed early in the application lifecycle, it may
+  /// report constraints with all dimensions set to zero.
+  ///
+  /// This value does not take into account any on-screen keyboards or other
+  /// system UI. If the constraints are tight, the [padding] and [viewInsets]
+  /// properties provide information about how much of each side of the view may
+  /// be obscured by system UI. If the consraints are loose, this information
```
Comment (ditman) [inline lib/ui/window.dart:158]:
```suggestion
  /// be obscured by system UI. If the constraints are loose, this information
```

File: lib/ui/window.dart [hunk 7]
```diff
@@ -324,25 +358,32 @@ class FlutterView {
   /// then obtain a [Scene] object, which you can display to the user via this
   /// [render] function.
   ///
+  /// If the view is configured with loose [physicalConstraints] (i.e.
+  /// [ViewConstraints.isTight] returns false) a `size` satisfying those
+  /// constraints must be provided. This method does not check that the provided
+  /// `size` actually meets the constraints (this should be done in a heigher
```
Comment (ditman) [inline lib/ui/window.dart:364]:
```suggestion
  /// `size` actually meets the constraints (this should be done in a higher
```

File: lib/web_ui/lib/platform_dispatcher.dart [hunk 8]
```diff
@@ -288,6 +288,25 @@ abstract class ViewPadding {
   }
 }
 
+abstract class ViewConstraints {
```
Comment (ditman) [inline lib/web_ui/lib/platform_dispatcher.dart:291]: I wonder if we can just show the same class coming from the normal `ui` package.

File: lib/web_ui/lib/src/engine/window.dart [hunk 9]
```diff
@@ -89,8 +89,9 @@ base class EngineFlutterView implements ui.FlutterView {
   }
 
   @override
-  void render(ui.Scene scene) {
+  void render(ui.Scene scene, {ui.Size? size}) {
     assert(!isDisposed, 'Trying to render a disposed EngineFlutterView.');
+    // TODO(goderbauer): Resepct the provided size when "physicalConstraints" are not allways tight. See TODO on "physicalConstraints".
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/window.dart:94]:
```suggestion
    // TODO(goderbauer): Respect the provided size when "physicalConstraints" are not always tight. See TODO on "physicalConstraints".
```

File: lib/web_ui/lib/src/engine/window.dart [hunk 10]
```diff
@@ -114,6 +115,10 @@ base class EngineFlutterView implements ui.FlutterView {
   late final PlatformViewMessageHandler platformViewMessageHandler =
       PlatformViewMessageHandler(platformViewsContainer: dom.platformViewsHost);
 
+  // TODO(goderbauer): Provide API to configure constraints. See also TODO in "render".
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/window.dart:118]: This will happen through the new JS Embedder API, when calling addView on an app.

**Review by ditman — APPROVED** [review]
LGTM!

(Here's [the changes I did](https://github.com/goderbauer/engine/compare/dynamic-view-sizing...ditman:flutter-engine:dynamic-view-sizing-web?expand=1) to get web support)

**Review by dkwingsmt — APPROVED** [review]
LGTM

File: lib/ui/window.dart (hunk 5 above)
Comment (goderbauer) [reply lib/ui/window.dart:149]: This is up to the embedders. If they want, they can offer an API to change this after View creation. If they do, `PlatformDispatcher.onMetricsChanged` needs to be invoked when it changes.

File: lib/ui/window.dart (hunk 3 above)
Comment (goderbauer) [reply lib/ui/window.dart:374]: This is essentially covered by https://github.com/flutter/flutter/issues/137073, I think. I left a comment to consider adding the assert when we fix that.

File: lib/web_ui/lib/platform_dispatcher.dart (hunk 8 above)
Comment (goderbauer) [reply lib/web_ui/lib/platform_dispatcher.dart:291]: I modeled this after ViewPadding above. But, yeah, I was also surprised that all this is repeated in web_ui...

File: lib/ui/window.dart (hunk 5 above)
Comment (ditman) [reply lib/ui/window.dart:149]: Fantastic!

File: lib/ui/platform_dispatcher.dart [hunk 11]
```diff
@@ -1981,6 +1981,113 @@ class ViewPadding {
...
+  /// Creates view constraints with the given constraints.
+  const ViewConstraints({
+    this.minWidth = 0.0,
+    this.maxWidth = double.infinity,
+    this.minHeight = 0.0,
+    this.maxHeight = double.infinity,
+  });
+
+  /// Creates view constraints that is respected only by the given size.
+  ViewConstraints.tight(Size size)
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply lib/ui/platform_dispatcher.dart:2011]: `size.width` and `size.height` are invalid constant values, according to Dart (In theory, they could be getters that do some crazy computation using information only available at runtime.). Since they are used in the Initializer list below, the constructor cannot be const, sadly.

(Same applies to the existing `BoxConstraints.tight` constructor)

[the remark this answers is not in the document]
Comment (gspencergoog) [reply lib/ui/platform_dispatcher.dart:2011]: Ahh, right, makes sense. Too bad there's no way to make a final property not be overridable with a getter.

File: lib/ui/window.dart [hunk 12]
```diff
@@ -138,8 +138,42 @@ class FlutterView {
...
+  /// constraints are typically used by an UI framework as the input for its
+  /// layout algorithm to determine an approrpiate size for the view. To size
+  /// the view, the selected size must be provided to the [render] method and it
+  /// must satisfy the constraints.
+  ///
+  /// When this changes, [PlatformDispatcher.onMetricsChanged] is called.
+  ///
+  /// At startup, the constraints for the view may not be known before Dart code
+  /// runs. If this value is observed early in the application lifecycle, it may
+  /// report constraints with all dimensions set to zero.
```
Comment (dkwingsmt) [inline lib/ui/window.dart:153]: Does "constraints with all dimensions set to zero" mean the resulting size must be 0? I don't know if this sentence needs to be said, because the sentence alone feels important on the surface but it's really just a temporary state. And the constraints are less like "known" but more like "set" by the engine.

File: lib/ui/window.dart [hunk 13]
```diff
@@ -138,8 +138,42 @@ class FlutterView {
...
+  ///
+  /// This value does not take into account any on-screen keyboards or other
+  /// system UI. If the constraints are tight, the [padding] and [viewInsets]
```
Comment (dkwingsmt) [inline lib/ui/window.dart:156]: Does "not take into account" mean the size include or exclude these system UI? I think include/exclude is clearer.

File: lib/ui/window.dart [hunk 14]
```diff
@@ -138,8 +138,42 @@ class FlutterView {
...
+  /// properties provide information about how much of each side of the view may
+  /// be obscured by system UI. If the constraints are loose, this information
+  /// is not known upfront.
```
Comment (dkwingsmt) [inline lib/ui/window.dart:159]: What do you mean by "this information is not known"? Shall we just assert that only tight constraints can have non-zero padding and insets?

File: lib/ui/window.dart (hunk 12 above)
Comment (goderbauer) [reply lib/ui/window.dart:153]: Yes, it means the only size allowed will be zero. It matches what was previously documented on `physicalSize`.

File: lib/ui/window.dart (hunk 14 above)
Comment (goderbauer) [reply lib/ui/window.dart:159]: This is basically describing platform limitations. No platform currently makes it possible to know this information upfront, but in theory a platform could provide it.
